### ISSUE TITLE
Question: Why is Query not a monad

### ISSUE THREAD

**Issue by tomjaguarpaw** [nikita-volkov/hasql#70]
Why is Query not given a monad instance.  The implementation is a monad.  Is there some technical reason the full generality of the monad cannot be exposed?

**Comment by nikita-volkov** [comment]
It's not a monad, it's an arrow.

**Comment by tomjaguarpaw** [comment]
OK, let me rephrase.  Why not expose an API like this:

````haskell
newtype Query' a =
  Query (ReaderT Connection.Connection (EitherT Decoders.Results.Error IO) a)
  deriving (Monad)
````

**Comment by nikita-volkov** [comment]
That API is `Session` :)

**Comment by tomjaguarpaw** [comment]
OK, I now see that `Query` has an `ArrowApply` instance, so I need to change my question again.

Since `ArrowApply` and `Monad` are isomorphic, why not just use the `Monad` version?  `Monad`s are much easier to work with than `Arrow`s.

**Comment by nikita-volkov** [comment]
Query was never intended to be either Monad or Arrow even. Originally it was [just a product, grouping other abstractions](https://github.com/nikita-volkov/hasql/commit/838bc8cb8ea6916febf583c3986d3ad90cb84340#diff-2d6a28a32344c717e92f273612d4f796L128). The maximum instance that it had was for Profunctor. It just happened so that at some point I realised that I could extend it to Arrow, so I did that. That instance actually achieves nothing that can't be done with Session, so it's likely to be removed later.

Query was originally intended to be an isolated abstraction over a single statement, which makes the user focus on that specific task. The API indulges users to have a separate module containing just isolated queries. With that said, I consider the introduction of the Arrow interface my design mistake. I should have let it be just a Profunctor. Hence that is a likely change in the future releases.

I must notice though that lately I've been thinking about a radical API update, which might get rid of Query altogether, living just the Session abstraction level. I haven't deeply studied the consequences though and whether that would be beneficial for the API. But that change is a possibility. I sense that it might be what you're talking about here as well.
